### ISSUE TITLE
[HOLD for payment 2023-05-23] [$1000] auto link (google.com) following by email domain @google.com not parsed.

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#18443]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. Open the app
2. Open any report
3. type this message @google.com [google.com](http://google.com/) [google.com](http://google.com/) and send message

## Expected Result:
message part email domain should not parsed but the the following auto link should parsed

## Actual Result:
all auto link following email domain not parsed

## Workaround:
Can the user still use Expensify without this being fixed? Have you informed them of the workaround?

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android / native
- [x] Android / Chrome
- [x] iOS / native
- [x] iOS / Safari
- [x] MacOS / Chrome / Safari
- [x] MacOS / Desktop

**Version Number:** 1.3.10-2
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

https://user-images.githubusercontent.com/43996225/236322393-afb34142-f8bc-4477-b525-c7ba71f412ba.mov

<img width="960" alt="Untitled" src="https://user-images.githubusercontent.com/43996225/236322418-38d22953-f95d-424d-b7aa-dde8e7f3ab9f.png">

**Expensify/Expensify Issue URL:**
**Issue reported by:** @ahmedGaber93 
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1683202864364059

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01804b7ba7971cb0b2</li>
        <li>Upwork Job ID: 1654612776396419072</li>
        <li>Last Price Increase: 2023-05-05</li>
    </ul>
</details>

**Comment by ahmedGaber93** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
 auto link (google.com) following by email domain @google.com not parsed. 
### What is the root cause of that problem?
[here](https://github.com/Expensify/expensify-common/blob/3cdaa947fe77016206c15e523017cd50678f2359/lib/ExpensiMark.js#L366) we we use `abort` inside loop, and when is set to true inside loop in the first time, it will still true for all next times in loop, see code below for explanation
```javascript
let abort = false; // variable scope out of while
let i = 0;
while (i < 5) {
    // we need to abort fist time only
    // but because abort default value out of while scope if it set to true in first time, it still true for all next times inside loop
    if (i === 0){
        abort = true;
    }
    
    if(abort) console.log(i);
    i ++;
}

//log result 0 1 2 3 4
```
### What changes do you think we should make in order to solve the problem?
set abort inside the while loop in this [function](https://github.com/Expensify/expensify-common/blob/3cdaa947fe77016206c15e523017cd50678f2359/lib/ExpensiMark.js#L366)
```javascript
let abort = false;
while () {
    // set default value inside the loop to reset every time loop
    // or define it in while scope let abort = false;
    abort = false; 
    if (some case){
        abort = true;
    }
}
```
### What alternative solutions did you explore? (Optional)

**Comment by sophiepintoraetz** [comment]
Yup - I'm able to replicate!
![image](https://user-images.githubusercontent.com/42940078/236366381-843b8f66-02fd-43e6-981b-fbb9f149dd5e.png)

**Comment by kushu7** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Its related to _ExpensiMark_ util from  _**expensify-common**_ library. 
https://github.com/Expensify/expensify-common/blob/main/lib/ExpensiMark.js#L392
when its check for urls in text in _modifyTextForUrlLinks_ function. its mark abort as true when its try to skip emails, after that its not set back it to default `abort= false` at loop startup. thats why we can see this type of behaviour.

### What changes do you think we should make in order to solve the problem?
We can set abort= true in loop. it will fix the issue.
 ```
   modifyTextForUrlLinks(regex, textToCheck, replacement) {
        let match = regex.exec(textToCheck);
        let replacedText = '';
        let startIndex = 0;
        

        while (match !== null) {
            let abort = false;
            // we want to avoid matching email address domains
            if ((match.index !== 0) && (textToCheck[match.index - 1] === '@')) {
                abort = true;
            }
          .....
        }
......
}

```
### What alternative solutions did you explore? (Optional)
No

**Comment by sophiepintoraetz** [comment]
All right, Melvin - chill out.

**Comment by sophiepintoraetz** [comment]
@parasharrajat - are you able to get to this? 🙏

**Comment by francoisl** [comment]
Yes looks good to me, let's go with your proposal @ahmedGaber93.

**Comment by ahmedGaber93** [comment]
@parasharrajat [PR](https://github.com/Expensify/expensify-common/pull/534) is now ready for review.

**Comment by ahmedGaber93** [comment]
@parasharrajat @francoisl [PR](https://github.com/Expensify/App/pull/18804) is now ready for review.

**Comment by sophiepintoraetz** [comment]
Shoot, great hustle here - so I think this is subject to the 7 day regression period, right?

**Comment by sophiepintoraetz** [comment]
I think we're just waiting for this to hit Prod.

**Comment by sophiepintoraetz** [comment]
@johncschuster - this is just awaiting payment tomorrow (provided there are no regressions). I checked your calendar and it said you'd be online for a bit so hoping you can get to this?

I've summarised the payment amounts [here](https://github.com/Expensify/App/issues/18443#issuecomment-1551714086) but just waiting on @parasharrajat to complete the [review checklist](https://github.com/Expensify/App/issues/18443#issuecomment-1550461566) here before payment.

**Comment by francoisl** [comment]
@parasharrajat can you fill out the checklist [here](https://github.com/Expensify/App/issues/18443#issuecomment-1550461566) please so that we can complete the payments and close this issue? Thanks in advance!

**Comment by francoisl** [comment]
Thanks @parasharrajat.

👍 for the proposed regression steps (though maybe we can use `expensify.com` instead of `google.com` :trollface: 😄)


@johncschuster - all yours for [payment](https://github.com/Expensify/App/issues/18443#issuecomment-1551714086) and opening the [regression steps](https://github.com/Expensify/App/issues/18443#issuecomment-1564634008) issue. Thanks!

**Comment by sophiepintoraetz** [comment]
Offers sent!

**Comment by ahmedGaber93** [comment]
Hi @tienifr, this is the first time I saw this part of your proposal, and regard to how I find this issue? When I'm working in this issue #17772 (Avoid parsing nested link), I add proposal (see first copy of my proposal [here](https://github.com/Expensify/App/issues/17772#issuecomment-1530476469)) based on `abort` variable like this, and I fall in this issue.
```javascript
let abort = false;
let i = 0;
while (...) {
    ...
    if (replacedText.lastIndexOf('<a') > replacedText.lastIndexOf('</a')){
       // this auto link match is nested 
        abort = true;
    }
}
```
If I saw this part of your proposal and I knew the `abort` issue, how can I add a proposal based on `abort` and I fall in the same issue?

So be completely sure that if I had seen this part of your proposal, I wouldn't have fell in the same issue (as I described above), and also I wouldn't have reported it here.
